File: src/types/graphData.ts
import type { TableWindowFilter } from "./data";

export interface GraphFieldBinding {
  role: string;
  column: string;
}

export interface GraphElementRequest {
  kind: string;
  summaryStat: string;
}

export type GraphSampling =
  | { mode: "full" }
  | { mode: "sample"; size: number; seed: number };

export interface GraphViewport {
  width: number;
  height: number;
}

export interface GraphDataRequest {
  requestId: string;
  datasetId: string;
  generation: number;
  fields: GraphFieldBinding[];
  filters: TableWindowFilter[];
  elements: GraphElementRequest[];
  sampling: GraphSampling;
  viewport: GraphViewport;
}

export type GraphPayloadType = "f64" | "u32" | "i64" | "u8";

export interface GraphTypedSliceDescriptor {
  type: GraphPayloadType;
  offset: number;
  byteLength: number;
}

export type GraphAxisEncoding = "numeric" | "categorical";

export interface GraphChunkHeader {
  requestId: string;
  generation: number;
  chunkIndex: number;
  rowOffset: number;
  rowCount: number;
  sourceRows: number;
  processedRows: number;
  dictionaries: Record<string, readonly string[]>;
  validityRanges: Record<string, GraphTypedSliceDescriptor>;
  xValues: GraphTypedSliceDescriptor;
  yValues: GraphTypedSliceDescriptor;
  rowIds: GraphTypedSliceDescriptor;
  groupCodes?: GraphTypedSliceDescriptor;
  sizeValues?: GraphTypedSliceDescriptor;
  xEncoding: GraphAxisEncoding;
  finalChunk: boolean;
}

export interface GraphDataCompletion {
  requestId: string;
  datasetId: string;
  generation: number;
  sourceRows: number;
  processedRows: number;
  chunksSent: number;
  cancelled: boolean;
}

export interface GraphChunkMessage {
  header: GraphChunkHeader;
  payload: ArrayBuffer;
}

export type GraphAggregatePacket =
  | { kind: "histogram"; payload: unknown }
  | { kind: "heatmap"; payload: unknown }
  | { kind: "boxPlot"; payload: unknown }
  | { kind: "summary"; payload: unknown };

export interface DecodedRawPointChunk {
  chunkIndex: number;
  rowOffset: number;
  rowCount: number;
  xValues: Float64Array | Uint32Array;
  yValues: Float64Array;
  rowIds: BigInt64Array;
  groupCodes?: Uint32Array;
  sizeValues?: Float64Array;
  validity: Record<string, Uint8Array>;
}

export interface GraphDataFrame {
  requestId: string;
  datasetId: string;
  generation: number;
  sourceRows: number;
  processedRows: number;
  sampling: GraphSampling;
  dictionaries: Record<string, readonly string[]>;
  extents: Record<string, { min: number; max: number }>;
  rawChunks: readonly DecodedRawPointChunk[];
  aggregates: readonly GraphAggregatePacket[];
}

export interface DecodedGraphChunk extends DecodedRawPointChunk {
  requestId: string;
  generation: number;
  sourceRows: number;
  processedRows: number;
  dictionaries: Record<string, readonly string[]>;
  xEncoding: GraphAxisEncoding;
  finalChunk: boolean;
}

export class GraphPayloadError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GraphPayloadError";
  }
}

const TYPE_WIDTH: Record<GraphPayloadType, number> = {
  f64: 8,
  u32: 4,
  i64: 8,
  u8: 1,
};

function validateDescriptor(
  descriptor: GraphTypedSliceDescriptor,
  label: string,
  payloadLength: number,
): [start: number, end: number] {
  if (!Number.isInteger(descriptor.offset) || descriptor.offset < 0) {
    throw new GraphPayloadError(`${label} offset must be a non-negative integer`);
  }
  if (!Number.isInteger(descriptor.byteLength) || descriptor.byteLength < 0) {
    throw new GraphPayloadError(`${label} byteLength must be a non-negative integer`);
  }
  if (descriptor.offset % 8 !== 0) {
    throw new GraphPayloadError(`${label} offset ${descriptor.offset} is not 8-byte aligned`);
  }

  const width = TYPE_WIDTH[descriptor.type];
  if (descriptor.offset % width !== 0) {
    throw new GraphPayloadError(`${label} offset ${descriptor.offset} is misaligned for ${descriptor.type}`);
  }
  if (descriptor.byteLength % width !== 0) {
    throw new GraphPayloadError(`${label} byteLength ${descriptor.byteLength} is invalid for ${descriptor.type}`);
  }

  const end = descriptor.offset + descriptor.byteLength;
  if (!Number.isSafeInteger(end) || end > payloadLength) {
    throw new GraphPayloadError(`${label} range exceeds payload length`);
  }

  return [descriptor.offset, end];
}

function assertType(
  descriptor: GraphTypedSliceDescriptor,
  expected: GraphPayloadType | readonly GraphPayloadType[],
  label: string,
): void {
  const allowed = Array.isArray(expected) ? expected : [expected];
  if (!allowed.includes(descriptor.type)) {
    throw new GraphPayloadError(`${label} expected ${allowed.join("|")} but got ${descriptor.type}`);
  }
}

export function decodeGraphPayload(
  header: GraphChunkHeader,
  payload: ArrayBuffer,
): DecodedGraphChunk {
  assertType(header.xValues, ["f64", "u32"], "xValues");
  assertType(header.yValues, "f64", "yValues");
  assertType(header.rowIds, "i64", "rowIds");

  if (header.groupCodes) {
    assertType(header.groupCodes, "u32", "groupCodes");
  }
  if (header.sizeValues) {
    assertType(header.sizeValues, "f64", "sizeValues");
  }

  const ranges: Array<{ start: number; end: number; label: string }> = [];
  const payloadLength = payload.byteLength;

  const register = (descriptor: GraphTypedSliceDescriptor, label: string): void => {
    const [start, end] = validateDescriptor(descriptor, label, payloadLength);
    ranges.push({ start, end, label });
  };

  register(header.xValues, "xValues");
  register(header.yValues, "yValues");
  register(header.rowIds, "rowIds");
  if (header.groupCodes) {
    register(header.groupCodes, "groupCodes");
  }
  if (header.sizeValues) {
    register(header.sizeValues, "sizeValues");
  }

  for (const [key, descriptor] of Object.entries(header.validityRanges)) {
    assertType(descriptor, "u8", `validityRanges.${key}`);
    register(descriptor, `validityRanges.${key}`);
  }

  ranges.sort((a, b) => a.start - b.start);
  for (let index = 1; index < ranges.length; index += 1) {
    const previous = ranges[index - 1];
    const current = ranges[index];
    if (previous.end > current.start) {
      throw new GraphPayloadError(
        `slice overlap between ${previous.label} and ${current.label}`,
      );
    }
  }

  const xValues =
    header.xValues.type === "f64"
      ? new Float64Array(payload, header.xValues.offset, header.xValues.byteLength / 8)
      : new Uint32Array(payload, header.xValues.offset, header.xValues.byteLength / 4);
  const yValues = new Float64Array(payload, header.yValues.offset, header.yValues.byteLength / 8);
  const rowIds = new BigInt64Array(payload, header.rowIds.offset, header.rowIds.byteLength / 8);
  const groupCodes = header.groupCodes
    ? new Uint32Array(payload, header.groupCodes.offset, header.groupCodes.byteLength / 4)
    : undefined;
  const sizeValues = header.sizeValues
    ? new Float64Array(payload, header.sizeValues.offset, header.sizeValues.byteLength / 8)
    : undefined;

  const validity: Record<string, Uint8Array> = {};
  for (const [key, descriptor] of Object.entries(header.validityRanges)) {
    validity[key] = new Uint8Array(payload, descriptor.offset, descriptor.byteLength);
  }

  return {
    requestId: header.requestId,
    generation: header.generation,
    chunkIndex: header.chunkIndex,
    rowOffset: header.rowOffset,
    rowCount: header.rowCount,
    sourceRows: header.sourceRows,
    processedRows: header.processedRows,
    dictionaries: header.dictionaries,
    xEncoding: header.xEncoding,
    finalChunk: header.finalChunk,
    xValues,
    yValues,
    rowIds,
    groupCodes,
    sizeValues,
    validity,
  };
}

File: src-tauri/src/models/graph_data.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use super::table::TableWindowFilter;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GraphFieldBinding {
    pub role: String,
    pub column: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GraphElementRequest {
    pub kind: String,
    pub summary_stat: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GraphViewport {
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(tag = "mode", rename_all = "camelCase")]
pub enum GraphSampling {
    Full,
    Sample { size: usize, seed: u64 },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GraphDataRequest {
    pub request_id: String,
    pub dataset_id: String,
    pub generation: u64,
    pub fields: Vec<GraphFieldBinding>,
    pub filters: Vec<TableWindowFilter>,
    pub elements: Vec<GraphElementRequest>,
    pub sampling: GraphSampling,
    pub viewport: GraphViewport,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum GraphPayloadType {
    F64,
    U32,
    I64,
    U8,
}

impl GraphPayloadType {
    pub fn byte_width(&self) -> usize {
        match self {
            Self::F64 => 8,
            Self::U32 => 4,
            Self::I64 => 8,
            Self::U8 => 1,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GraphTypedSliceDescriptor {
    #[serde(rename = "type")]
    pub payload_type: GraphPayloadType,
    pub offset: usize,
    pub byte_length: usize,
}

impl GraphTypedSliceDescriptor {
    pub fn new(payload_type: GraphPayloadType, offset: usize, byte_length: usize) -> Self {
        Self {
            payload_type,
            offset,
            byte_length,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub enum GraphAxisEncoding {
    Numeric,
    Categorical,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GraphChunkHeader {
    pub request_id: String,
    pub generation: u64,
    pub chunk_index: u32,
    pub row_offset: u64,
    pub row_count: usize,
    pub source_rows: u64,
    pub processed_rows: u64,
    pub projected_columns: Vec<String>,
    pub dictionaries: BTreeMap<String, Vec<String>>,
    pub validity_ranges: BTreeMap<String, GraphTypedSliceDescriptor>,
    pub x_values: GraphTypedSliceDescriptor,
    pub y_values: GraphTypedSliceDescriptor,
    pub row_ids: GraphTypedSliceDescriptor,
    pub group_codes: Option<GraphTypedSliceDescriptor>,
    pub size_values: Option<GraphTypedSliceDescriptor>,
    pub x_encoding: GraphAxisEncoding,
    pub final_chunk: bool,
}

impl GraphChunkHeader {
    pub fn validate_layout(&self, payload_len: usize) -> Result<(), String> {
        let mut slices: Vec<&GraphTypedSliceDescriptor> = vec![
            &self.x_values,
            &self.y_values,
            &self.row_ids,
        ];

        if let Some(group_codes) = &self.group_codes {
            slices.push(group_codes);
        }
        if let Some(size_values) = &self.size_values {
            slices.push(size_values);
        }

        for descriptor in self.validity_ranges.values() {
            slices.push(descriptor);
        }

        for descriptor in &slices {
            if descriptor.offset % 8 != 0 {
                return Err(format!(
                    "slice offset {} is not 8-byte aligned",
                    descriptor.offset
                ));
            }

            let type_width = descriptor.payload_type.byte_width();
            if descriptor.offset % type_width != 0 {
                return Err(format!(
                    "slice offset {} is not aligned for {:?}",
                    descriptor.offset, descriptor.payload_type
                ));
            }
            if descriptor.byte_length % type_width != 0 {
                return Err(format!(
                    "slice byte length {} is not divisible by {:?}",
                    descriptor.byte_length, descriptor.payload_type
                ));
            }

            let end = descriptor
                .offset
                .checked_add(descriptor.byte_length)
                .ok_or_else(|| "slice range overflow".to_string())?;

            if end > payload_len {
                return Err(format!(
                    "slice range [{}..{}) exceeds payload length {}",
                    descriptor.offset, end, payload_len
                ));
            }
        }

        let mut ranges: Vec<(usize, usize)> = slices
            .iter()
            .map(|descriptor| (descriptor.offset, descriptor.offset + descriptor.byte_length))
            .collect();
        ranges.sort_unstable_by_key(|range| range.0);

        for pair in ranges.windows(2) {
            let prev = pair[0];
            let next = pair[1];
            if prev.1 > next.0 {
                return Err(format!(
                    "slice overlap detected between [{}..{}) and [{}..{})",
                    prev.0, prev.1, next.0, next.1
                ));
            }
        }

        Ok(())
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct GraphDataCompletion {
    pub request_id: String,
    pub dataset_id: String,
    pub generation: u64,
    pub source_rows: u64,
    pub processed_rows: u64,
    pub chunks_sent: u32,
    pub cancelled: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn graph_request_deserializes_with_camel_case_fields() {
        let request: GraphDataRequest = serde_json::from_value(serde_json::json!({
            "requestId": "req-1",
            "datasetId": "dataset-id",
            "generation": 7,
            "fields": [
                { "role": "x", "column": "region" },
                { "role": "y", "column": "cost" }
            ],
            "filters": [],
            "elements": [{ "kind": "points", "summaryStat": "none" }],
            "sampling": { "mode": "full" },
            "viewport": { "width": 1200, "height": 700 }
        }))
        .unwrap();

        assert_eq!(request.request_id, "req-1");
        assert!(matches!(request.sampling, GraphSampling::Full));
    }

    #[test]
    fn chunk_layout_requires_eight_byte_alignment_and_non_overlapping_slices() {
        let header = GraphChunkHeader {
            request_id: "req-1".into(),
            generation: 3,
            chunk_index: 0,
            row_offset: 0,
            row_count: 2,
            source_rows: 2,
            processed_rows: 2,
            projected_columns: vec!["_row_id".into(), "x".into(), "y".into()],
            dictionaries: std::collections::BTreeMap::new(),
            validity_ranges: std::collections::BTreeMap::from([
                (
                    "x".into(),
                    GraphTypedSliceDescriptor::new(GraphPayloadType::U8, 72, 2),
                ),
            ]),
            x_values: GraphTypedSliceDescriptor::new(GraphPayloadType::F64, 0, 16),
            y_values: GraphTypedSliceDescriptor::new(GraphPayloadType::F64, 16, 16),
            row_ids: GraphTypedSliceDescriptor::new(GraphPayloadType::I64, 32, 16),
            group_codes: Some(GraphTypedSliceDescriptor::new(GraphPayloadType::U32, 48, 8)),
            size_values: Some(GraphTypedSliceDescriptor::new(GraphPayloadType::F64, 56, 16)),
            x_encoding: GraphAxisEncoding::Categorical,
            final_chunk: true,
        };

        assert!(header.validate_layout(80).is_ok());

        let mut misaligned = header.clone();
        misaligned.x_values.offset = 4;
        assert!(misaligned.validate_layout(80).is_err());

        let mut overlapping = header.clone();
        overlapping.y_values.offset = 8;
        assert!(overlapping.validate_layout(80).is_err());
    }
}